draft-ietf-tls-tls13.md

```diff
@@ -1953,11 +1947,14 @@ The "extension_data" field of this extension contains a
 %%% Key Exchange Messages
 
        struct {
-           select (role) {
-               case client:
+           select (Handshake.msg_type) {
+               case client_hello:
                    KeyShareEntry client_shares<0..2^16-1>;
 
-               case server:
+               case hello_retry_request:
+                   NamedGroup selected_group;
+
+               case server_hello:
                    KeyShareEntry server_share;
            }
        } KeyShare;
```

Comment:
There's no disputing that it's a compatibility kludge to use two extensions for negotiation, but adding this kludge onto that kludge feels iffy. Using key_share to exchange a different structure from normal does not feel like the way to go here. I would much prefer less special-casing.

Using supported_groups makes more sense because the point of an HRR is to inform the client that the server doesn't support anything in the client's key_share, but does support something in its supported_groups. Just sending its full supported_groups extension could also work, if the client wishes to pick itself.